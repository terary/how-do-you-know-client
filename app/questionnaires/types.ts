export interface IQuestionFEMeta {
  isSkipped: boolean;
  userFlags: string;
  userSortPosition: number;
}

type TUserPromptType = "text" | "multimedia";
type TUserResponseType = "any-of" | "free-text-255" | "one-of-2" | "one-of-4";
type TTextItem = { labelText: string; value: string };
type TBooleanItem = { labelText: string; value: boolean };
type TMultimediaProperties = {
  mediaContentType: "audio/mpeg" | "video/mp4" | "link/youtube" | "link/image";
  height: number;
  width: number;
  url: string;
  specialInstructionText: string;
};

type TQuestionGeneric<
  // maybe these types shouldn't have default values
  UPromptType extends TUserPromptType = "text",
  UResponseType extends TUserResponseType = "free-text-255"
> = {
  userPromptType: UPromptType;
  userResponseType: UResponseType;
  questionId: string;
  userResponseHistory?: any[]; // for teaching tool this is a nice have. for exam this is a security risk (maybe).
  feMeta?: IQuestionFEMeta;
};

type TQuestionTextOneOf4 = TQuestionGeneric<"text", "one-of-4"> & {
  userPromptText: string;
  choices: TChoiceOption[];
  validAnswers: TTextItem[]; // sometimes submitting image or sound file - in which case it wont be multiple guess
  answerFodderId: string;
  userResponse: { value: string | null };
};

type TQuestionGenericUserPromptMultimedia = TQuestionGeneric<
  "multimedia",
  any
> & {
  links: TMultimediaProperties[];
};

// *tmc* - probably want to re-think this type inheritance
// we want generic userPrompt(Text|Multimedia) generic with user response.
// we do not want spaghetti inheritance
type TQuestionUserPromptText = TQuestionGeneric<"text", any> & {
  userPromptText: string;
};

type TQuestionTextText = TQuestionGeneric<"text", "free-text-255"> & {
  userPromptText: string;
  userResponse: { text: string | null };
};

type TQuestionTextTrueFalse = TQuestionGeneric<"text", "one-of-2"> & {
  userPromptText: string;
  validAnswers: TBooleanItem[]; // sometimes submitting image or sound file - in which case it wont be multiple guess
  userResponse: { value: "true" | "false" | null };
};
type TChoiceOption = {
  labelText: string;
  value: string;
};

type TQuestionTextAnyOf = TQuestionGeneric<"text", "any-of"> & {
  userPromptText: string;
  validAnswers: TTextItem[]; // sometimes submitting image or sound file - in which case it wont be multiple guess
  choices: TChoiceOption[];

  userResponse: { values: string[] | null };
};

// --------------

type TQuestionMultimediaText = TQuestionGeneric<
  "multimedia",
  "free-text-255"
> & {
  links: TMultimediaProperties[];
  instructionText: string;
  userResponse: { text: string | null };
};

type TQuestionMultimediaOneOf4 = TQuestionGeneric<"multimedia", "one-of-4"> & {
  links: TMultimediaProperties[];
  instructionText: string;
  choices: TChoiceOption[];

  validAnswers: TTextItem[]; // sometimes submitting image or sound file - in which case it wont be multiple guess
  answerFodderId: string;
  userResponse: { value: string | null };
};

type TQuestionMultimediaAnyOf = TQuestionGeneric<"multimedia", "any-of"> & {
  links: TMultimediaProperties[];
  instructionText: string;
  choices: TChoiceOption[];
  validAnswers: TTextItem[]; // sometimes submitting image or sound file - in which case it wont be multiple guess
  answerFodderId: string;
  userResponse: { values: string[] | null };
};

type TQuestionMultimediaTrueFalse = TQuestionGeneric<
  "multimedia",
  "one-of-2"
> & {
  links: TMultimediaProperties[];
  instructionText: string;
  validAnswers: TBooleanItem[]; // sometimes submitting image or sound file - in which case it wont be multiple guess
  userResponse: { value: "true" | "false" | null };
};

type TMultimediaLink = TMultimediaProperties;

type TQuestionAny =
  | TQuestionGenericUserPromptMultimedia
  | TQuestionTextAnyOf
  | TQuestionTextOneOf4
  | TQuestionTextText
  | TQuestionTextTrueFalse
  | TQuestionMultimediaAnyOf
  | TQuestionMultimediaOneOf4
  | TQuestionMultimediaText
  | TQuestionMultimediaTrueFalse;

type TQuestionUserResponseText = TQuestionTextText | TQuestionMultimediaText;
type TQuestionUserResponseAnyOf = TQuestionTextAnyOf | TQuestionMultimediaAnyOf;
// TQuestionUserResponseText
type TQuestionUserResponseOneOf4 =
  | TQuestionTextOneOf4
  | TQuestionMultimediaOneOf4;

export type {
  TChoiceOption,
  TMultimediaLink,
  TQuestionAny,
  TQuestionUserPromptText,
  TQuestionUserResponseAnyOf,
  TQuestionUserResponseOneOf4,
  TQuestionUserResponseText,
  TQuestionMultimediaOneOf4,
  TQuestionTextOneOf4,
  TQuestionGenericUserPromptMultimedia,
  TUserResponseType,
};

export type TQuestionText = {
  userPromptType: "text";
  userPromptText: string;
};

export type TQuestionMultimedia = {
  userPromptType: "multimedia";
  instructionText: string;
  links: Array<{
    mediaContentType: "link/youtube" | "link/image";
    url: string;
    width: number;
    height: number;
    specialInstructionText: string;
  }>;
};
