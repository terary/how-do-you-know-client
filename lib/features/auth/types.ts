import { User } from "../../types";

export interface LoginCredentials {
  username: string;
  password: string;
}

export interface AuthState {
  user: User | null;
  isAuthenticated: boolean;
  token: string | null;
}

export interface LoginResponse {
  access_token: string;
  user: User;
}

export interface RegisterCredentials extends LoginCredentials {
  name: string;
  confirmPassword: string;
}
