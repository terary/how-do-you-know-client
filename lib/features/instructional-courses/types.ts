export type DayOfWeek =
  | "MONDAY"
  | "TUESDAY"
  | "WEDNESDAY"
  | "THURSDAY"
  | "FRIDAY"
  | "SATURDAY"
  | "SUNDAY";

export interface InstructionalCourse {
  id: string;
  name: string;
  description: string;
  start_date: string;
  finish_date: string;
  start_time_utc: string;
  duration_minutes: number;
  days_of_week: DayOfWeek[];
  institution_id: string;
  instructor_id: string;
  proctor_ids: string[];
  created_by: string;
  created_at: string;
  updated_at: string;
}

export interface CreateInstructionalCourseDto {
  name: string;
  description: string;
  start_date: string;
  finish_date: string;
  start_time_utc: string;
  duration_minutes: number;
  days_of_week: DayOfWeek[];
  institution_id: string;
  instructor_id: string;
  proctor_ids: string[];
}

export interface UpdateInstructionalCourseDto {
  name?: string;
  description?: string;
  start_date?: string;
  finish_date?: string;
  start_time_utc?: string;
  duration_minutes?: number;
  days_of_week?: DayOfWeek[];
  institution_id?: string;
  instructor_id?: string;
  proctor_ids?: string[];
}
