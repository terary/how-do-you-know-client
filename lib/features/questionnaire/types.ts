import { User } from "../../types";

export interface Question {
  id: string;
  text: string;
  type: "text" | "multiple_choice" | "boolean";
  options?: string[];
}

export interface Questionnaire {
  id: string;
  title: string;
  description: string;
  questions: Question[];
  createdBy: User;
  createdAt: string;
}

export interface UserResponse {
  questionId: string;
  answer: string | boolean | number;
  timestamp: string;
}
