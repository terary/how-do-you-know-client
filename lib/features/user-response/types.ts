import type { TUserResponseType } from "@/app/questionnaires/types";

export type TUserResponse = {
  questionId: string;
  systemAcceptTimeUtc?: number; // this is backend generated so it's not available when FE creates the update request
  systemUserResponseId?: string; // this is backend generated so it's not available when FE creates the update request

  userResponseType: TUserResponseType;
  userResponse: {
    text: string;
  };
};
