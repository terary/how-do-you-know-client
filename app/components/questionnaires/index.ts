export * from "./QuestionList";
export * from "./AcceptedAnswer";
export * from "./QuestionAny";
export * from "./UserResponseContainer";
