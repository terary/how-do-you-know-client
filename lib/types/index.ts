// Common types
export interface User {
  id: string;
  email: string;
  name: string;
}

export interface ApiResponse<T> {
  data: T;
  status: number;
  message?: string;
}

// Re-export feature types
export * from "../features/questionnaire/types";
export * from "../features/auth/types";
