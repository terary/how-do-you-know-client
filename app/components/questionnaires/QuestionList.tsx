import { FC } from "react";
import { useSelector } from "react-redux";
import { RootState } from "@/lib/store";
import { QuestionAny } from "./QuestionAny";
import { Card, CardContent } from "@mui/material";
import { UserResponseContainer } from "./UserResponseContainer";
import { TQuestionAny } from "@/app/questionnaires/types";

export const QuestionList: FC = () => {
  const questionMap = useSelector(
    (state: RootState) => state.userResponseUI.questionMap
  );

  return (
    <div>
      {Object.values(questionMap || {}).map((question) => (
        <Card
          key={question.questionId}
          variant="elevation"
          elevation={5}
          sx={{ margin: "10px" }}
        >
          <CardContent>
            <QuestionAny question={question} />
            <UserResponseContainer question={question as TQuestionAny} />
          </CardContent>
        </Card>
      ))}
    </div>
  );
};
