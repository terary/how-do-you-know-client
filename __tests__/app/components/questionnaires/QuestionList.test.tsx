import { render } from "../../../test-utils";
import { QuestionList } from "@/app/components/questionnaires/QuestionList";
import { TQuestionAny } from "@/app/questionnaires/types";

describe("QuestionList", () => {
  const mockQuestions = {
    "1": {
      questionId: "1",
      userPromptType: "text",
      userPromptText: "First question",
      userResponseType: "free-text-255",
      userResponseHistory: [],
    },
    "2": {
      questionId: "2",
      userPromptType: "text",
      userPromptText: "Second question",
      userResponseType: "one-of-4",
      userResponseHistory: [],
      choices: [
        { labelText: "Option A", value: "A" },
        { labelText: "Option B", value: "B" },
      ],
    },
  };

  const basePreloadedState = {
    userResponseUI: {
      questionMap: mockQuestions,
      draftResponses: {},
      isEditing: false,
    },
    questionFilter: {
      showSkipped: false,
      tagFilter: "",
    },
  };

  it("renders list of questions from redux store", () => {
    const { getByText } = render(<QuestionList />, {
      preloadedState: basePreloadedState,
    });
    expect(getByText("First question")).toBeInTheDocument();
    expect(getByText("Second question")).toBeInTheDocument();
  });

  it("renders empty when no questions in store", () => {
    const preloadedState = {
      ...basePreloadedState,
      userResponseUI: {
        ...basePreloadedState.userResponseUI,
        questionMap: {},
      },
    };

    const { container } = render(<QuestionList />, { preloadedState });
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders nothing when questionMap is undefined", () => {
    const preloadedState = {
      ...basePreloadedState,
      userResponseUI: {
        ...basePreloadedState.userResponseUI,
        questionMap: undefined,
      },
    };

    const { container } = render(<QuestionList />, { preloadedState } as any);
    expect(container.firstChild).toBeDefined();
    expect(container.firstChild?.childNodes.length).toBe(1); // Just the AdvancedQuestionSortFilter
  });
});
