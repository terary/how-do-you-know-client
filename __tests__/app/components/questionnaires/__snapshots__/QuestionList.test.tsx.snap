// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionList renders empty when no questions in store 1`] = `<div />`;
